Better validate partial measurement output

efernandez opened the pull request:
* Show source if validation fails
  * Also changes from throwing/crashing to `ROS_ERROR`.

* Add precision arg for covariance validation. This allows to:
    * Relax the default precision when validating the covariance matrix is
      symmetric.
    * Print the covariance matrix with `Eigen::FullPrecision` when the
      symmetry test fails with `isApprox`, so we can see the magnitude of
      the error.

efernandez committed:
Add precision arg for covariance validation

This allows to:
* Relax the default precision when validating the covariance matrix is
  symmetric.
* Print the covariance matrix with `Eigen::FullPrecision` when the
  symmetry test fails with `isApprox`, so we can see the magnitude of
  the error.

efernandez committed:
Show source if validation fails

Also changes from throwing/crashing to ROS_ERROR.

svwilliams commented on fuse_models/include/fuse_models/common/sensor_proc.h:
```diff
@@ -459,7 +471,16 @@ inline bool processDifferentialPoseWithCovariance(
+  {
+    ROS_ERROR_STREAM_THROTTLE(10.0, "Invalid partial differential pose measurement from '"
+                                        << source << "' source: " << ex.what());
+  }
```
I feel like if the validation fails, we should not be creating a constraint from it. In the same way that tf errors prevent the creation of a constraint. Thoughts?

```suggestion
    return false;
  }
```

efernandez replied:
The reason I still create constraints it's because I'm playing with https://github.com/clearpathrobotics/fuse/pull/7 and I still get cases where the covariance is slightly non-symmetric or non-PSD, but the constraint with those slightly bad covariances don't make the optimizer explode.

Maybe we could wait til EOW to see if I get things to actually pass this checks. I'm also planning to create an upstream PR for https://github.com/clearpathrobotics/fuse/pull/7 once I get things to work better.

ayrton04 replied:
Would an extra parameter work? Something like `permit_invalid`?

efernandez replied:
Actually, we could have a parameter to disable the checks. Something like `disable_checks` in Ceres solver.

The checks could be enabled by default, so the user can easily detect issues with the inputs and understand what's wrong with them. This would incur a small overhead.

Once things are known to work and be correct consistently, then `disable_checks: true` could be set, so we don't incur any overhead. If you're ok with something like this, I can add it.

efernandez replied:
@svwilliams What's your opinion on this?

svwilliams replied:
I like the "disable_checks" theory. With checks enabled, failed tests prevent constraints from being generated. And with checks disabled, the constraints are obviously generated all the time.

efernandez replied:
Add `disable_checks` param (defaults to `false`) to all sensor models and `validate` argument to all `sensor_proc.h` processing functions.

If `disable_checks == false`, then `validate == true` and all validation checks are evaluated. When a validation check fails, no constraint is created for the partial measurements involved.

If `disable_checks == true`, then `validate == false` and no validation check is evaluated. Therefore, constraints are created for all partial measurements.

efernandez commented on fuse_models/include/fuse_models/common/sensor_proc.h:
```diff
@@ -189,25 +189,28 @@ inline void populatePartialMeasurement(
   {
-    throw std::runtime_error("Non-positive-definite partial covariance matrix " +
+    throw std::runtime_error("Non-positive-definite partial covariance matrix\n" +
                              fuse_core::to_string(covariance_partial));
```
I'd like to change this to `fuse_core::to_string(covariance_partial, Eigen::FullPrecision)` as well.

And also print the `solver.eigenvalues()`.

Sounds good?

Mostly a heads up, so you wait for those changes before you merge this PR. :smiley:

svwilliams replied:
More information on failure is better than less. Go for it.

efernandez replied:
Added the additional information.

efernandez committed:
Add eigenvalues to non-PSD error check

And use `Eigen::FullPrecision` output.

efernandez committed:
Add disable_checks param to sensor models

And `validate` argument to `sensor_proc.h` processing functions, so the
validation checks can be enabled or disabled based on the value of the
`disable_checks` param.
